### PR TITLE
Add Alpine 3.21 Dockerfile

### PR DESCRIPTION
**Pull request by mthalman**
Related to https://github.com/dotnet/core/issues/9577

I didn't continue the pattern of defining both `alpine/3.21/amd64/Dockerfile` and `alpine/3.21/withnode/amd64/Dockerfile` as has been done for previous Alpine versions, because the non-Node Dockerfile is no longer being used by any builds. So I've consolidated it into a single Dockerfile.

This also only defines an architecture-specific tag. Previous Alpine versions had a `alpine-<version>-withnode` but that doesn't follow the new tagging guidelines. So it no longer provides that tag for backwards compatibility. Related: https://github.com/dotnet/dotnet-buildtools-prereqs-docker/pull/1262/files#diff-0983c53ccd3b14ab97c1b4d7dec524efc89e51c56ebc90d96ce1b8ba5d4fa100

### CODE REVIEW

File: src/alpine/3.21/withnode/amd64/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    && LATEST_TAG=$(curl -L https://api.github.com/repos/powershell/powershell/releases/latest | jq -r '.tag_name') \
+    && curl -L https://github.com/PowerShell/PowerShell/releases/download/$LATEST_TAG/powershell-${LATEST_TAG#*v}-linux-musl-x64.tar.gz -o /tmp/powershell.tar.gz \
+    && mkdir -p /opt/microsoft/powershell \
+    && tar zxf /tmp/powershell.tar.gz -C /opt/microsoft/powershell \
+    && chmod +x /opt/microsoft/powershell/pwsh \
+    && ln -s /opt/microsoft/powershell/pwsh /usr/bin/pwsh \
+    && rm -f /tmp/powershell.tar.gz
+
+# Install azurecli from PIP
+RUN azureEnv="/usr/local/share/azure-cli-env" && \
```
Comment (MichaelSimons) [inline src/alpine/3.21/withnode/amd64/Dockerfile:70]: Can you make this Dockerfile consistent with respect to the '&&` formatting?  The previous section uses a prefix pattern where this section uses a suffix pattern.  I have been trying to standardize around the prefix pattern.

File: src/alpine/3.21/withnode/amd64/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    "$azureEnv/bin/python" -m pip install azure-cli && \
+    ln -s "$azureEnv/bin/az" /usr/local/bin/az
+
+# Add label for bring your own node in azure devops
+LABEL "com.azure.dev.pipelines.agent.handler.node.path"="/usr/bin/node"
+
+ENV NO_UPDATE_NOTIFIER=true
+
+# Set node as a default command
+CMD [ "node" ]
```
Comment (MichaelSimons) [inline src/alpine/3.21/withnode/amd64/Dockerfile:82]: What requires this?   It feels strange to have this set for a general build container.

Comment (mthalman) [reply src/alpine/3.21/withnode/amd64/Dockerfile:82]: @wtgodbe - Does your infrastructure require this?

File: src/alpine/3.21/withnode/amd64/Dockerfile (hunk 1 above)
Comment (mthalman) [reply src/alpine/3.21/withnode/amd64/Dockerfile]: Fixed in https://github.com/dotnet/dotnet-buildtools-prereqs-docker/pull/1284/commits/01d04470c6dcafa56c876a9f6fc78c68d6879228

**Comment by richlander** [comment]
I'd like to understand why this non-helix Dockerfile is needed at all. As I mentioned on the Debian 12 PR, we need to understand why anyone needs non-Azure Linux builder images.

**Comment by richlander** [comment]
When you say "validation", does that involve building the code or just testing it? And why does Alpine have a "WithNode" variant and the other distros do not?

My goal is to (A) ensure we understand the model, and (B) simplify it.

**Comment by richlander** [comment]
I propose that we switch this PR to be the same as the other images you need, like the Ubuntu one, with the same naming scheme and the same components. I assume you are not using it with Azure DevOps so no node is needed.

We can then figure out what the use case is for repos with the `WithNode` image references. They will definitely pop up.

**Comment by mthalman** [comment]
Node was initially added to images used by source build here: https://github.com/dotnet/dotnet-buildtools-prereqs-docker/pull/949

That requirement has since gone away but that may end up being temporary as we will likely need a way to build Node sources in the future. Most of the images used by source build still have Node in them. Interestingly Fedora had them removed by @MichaelSimons in https://github.com/dotnet/dotnet-buildtools-prereqs-docker/pull/1226, but I don't think that was intentional.

**Comment by richlander** [comment]
That's fine. Let's start by getting rid of the naming scheme.

**Comment by MichaelSimons** [comment]
> Interestingly Fedora had them removed by @MichaelSimons in https://github.com/dotnet/dotnet-buildtools-prereqs-docker/pull/1226, but I don't think that was intentional.

It was intentional as aspnet just added them for source-build.  Cleaning this up wholesale and eliminating the special naming is good.

**Review by MichaelSimons — APPROVED** [review]

**Review by richlander — APPROVED** [review]

**Comment by ellahathaway** [comment]
Having node is a requirement for containerizing jobs using non-glibc-based containers: [Container jobs in YAML - Azure Pipelines | Microsoft Learn](https://learn.microsoft.com/en-us/azure/devops/pipelines/process/container-phases?view=azure-devops&tabs=linux#nonglibc-based-containers)

Removing node blocks my ability to move forward with https://github.com/dotnet/source-build/issues/4547

**Comment by MichaelSimons** [comment]
> Having node is a requirement for containerizing jobs using non-glibc-based containers: [Container jobs in YAML - Azure Pipelines | Microsoft Learn](https://learn.microsoft.com/en-us/azure/devops/pipelines/process/container-phases?view=azure-devops&tabs=linux#nonglibc-based-containers)

That is reason enough to add it back IMO.

File: src/alpine/3.21/withnode/amd64/Dockerfile (hunk 2 above)
Comment (wtgodbe) [reply src/alpine/3.21/withnode/amd64/Dockerfile]: Yes, historically we've relied on it

**Pull request merged**